src/js/components/returns/inbound/AddItemsPage.jsx

```diff
@@ -314,61 +343,82 @@ class AddItemsPage extends Component {
+      if (data.inventoryItem && data.inventoryItem.expirationDate !== it.expirationDate) {
+        isSomeItemsDontMatchExpirationDate = true;
+        break;
+      }
+    }
+
+    // After find at least a single instance where expiration date we are trying to save
+    // does not match the existing inventoryItem expiration date, we want to inform the user
+    // that certain updates to th expiration date in the system will be performed
+    if (isSomeItemsDontMatchExpirationDate) {
+      const isConfirmed = await this.confirmExpirationDateSave();
+      if (!isConfirmed) {
+        return Promise.reject();
+      }
+    }
+
+    return this.saveStockTransfer(returnItems, status);
   }
 
-  saveAndExit(formValues) {
-    const errors = this.validate(formValues).returnItems;
-    if (errors.length && errors.every(obj => typeof obj === 'object' && _.isEmpty(obj))) {
-      this.saveStockTransferInCurrentStep(formValues.returnItems)
-        .then(() => {
-          window.location = STOCK_MOVEMENT_URL.show(this.props.match.params.inboundReturnId);
-        });
-    } else {
-      confirmAlert({
-        title: this.props.translate('react.inboundReturns.confirmSave.label', 'Confirm save'),
-        message: this.props.translate(
-          'react.inboundReturns.confirmExit.message',
-          'Validation errors occurred. Are you sure you want to exit and lose unsaved data?',
-        ),
-        buttons: [
-          {
-            label: this.props.translate('react.default.yes.label', 'Yes'),
-            onClick: () => { window.location = STOCK_MOVEMENT_URL.show(this.props.match.params.inboundReturnId); },
-          },
-          {
-            label: this.props.translate('react.default.no.label', 'No'),
-          },
-        ],
+  async save(formValues) {
+    this.saveStockTransferInCurrentStep(formValues)
+      .then(() => {
+        Alert.success(
+          this.props.translate(
+            'react.inboundReturns.alert.saveSuccess.label',
+            'Changes saved successfully',
+          ),
+          { timeout: 3000 },
+        );
       });
-    }
   }
 
-  saveItems(returnItems) {
-    this.props.showSpinner();
+  async saveAndExit(formValues) {
+    const errors = this.validate(formValues).returnItems;
+    const hasErrors = errors.length && errors.some((obj) => typeof obj === 'object' && !_.isEmpty(obj));
 
-    this.saveStockTransferInCurrentStep(returnItems)
-      .then(() => {
-        this.props.hideSpinner();
-        Alert.success(this.props.translate('react.inboundReturns.alert.saveSuccess.label', 'Changes saved successfully'), { timeout: 3000 });
-      })
-      .catch(() => this.props.hideSpinner());
+    if (hasErrors) {
+      const isConfirmed = await this.confirmValidationErrorOnExit();
+      if (!isConfirmed) {
+        return;
+      }
+    } else {
+      try {
+        await this.saveStockTransferInCurrentStep(formValues);
+      } catch (error) {
+        return;
```

Comment:
why do we actually want to do nothing when an error happens? Shouldn't we let the error handler handle the error and display it to a user?

Reply:
the error will be displayed to the user, because when using our custom `apiClient` we have a  response interceptor that displays an alert when we get a server error.

I am returning here to prevent the function going further, because the last step is redirecting to stock movement show page and if we have an error we don;t want to do that.

Reply:
@drodzewicz ok. I thought it would work like in Java (if we use the `catch`, we'd override the error handler and not let it handle the error).

Reply:
The difference between our grails error handler and frontend axios error handler is the order when they are called.
In case of grails, the global error handler is called as the last method, so when we catch a method manually it does not propagate to the global error handler.
For our custom `apiClient` on frontend we set the error handler as a response interceptor which is always called first. It handles the error and then returns a rejected promise.